### PR TITLE
[Inductor Intel GPU backend Upstream] Add Inductor Intel GPU backend.

### PR DESCRIPTION
**Pull request by etaf**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #122866
* __->__ #121895

As the design in RFC https://github.com/pytorch/pytorch/issues/114856, this PR implemented Intel GPU Inductor backend by:
- Reuse WrapperCodegen and TritonScheduling for python wrapper and kernel code generation. And implenented device-specific code generation in XPUDeviceOpOverrides
- Reuse fx_pass, lowering, codecache, triton kernel auto-tuning, and compilation.

For the test case, this PR provided test/inductor/test_xpu_basic.py for basic inductor backend functionality testing.
We'll reuse all the existing Inductor test case in the next PR.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/_inductor/lowering.py [hunk 1]
```diff
@@ -3325,7 +3325,8 @@ def scatter_fallback(
         reduce not in {None, reduce_ty}
         or (
             isinstance(src, TensorBox)
-            and src.get_device().type == torch.device("cuda").type
+            and src.get_device().type
+            in [torch.device("cuda").type, torch.device("xpu").type]
```
Comment (EikanWang) [inline torch/_inductor/lowering.py:3329]:
```suggestion
            and src.get_device().type in ["cuda", "xpu"]
```

**Review by EikanWang — CHANGES_REQUESTED** [review]

File: test/inductor/test_xpu_basic.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+        self.common(fn, (torch.rand(2, 3, 16, 16), torch.rand(2, 3, 16, 16)))
+
+
+if __name__ == "__main__":
+    from torch._dynamo.test_case import run_tests
+    from torch.testing._internal.inductor_utils import HAS_XPU
+
+    if HAS_XPU:
+        run_tests(needs="filelock")
```
Comment (EikanWang) [inline test/inductor/test_xpu_basic.py:49]: @etaf , please check if we can reuse some Inductor test cases to verify the functionality.

Comment (etaf) [reply test/inductor/test_xpu_basic.py:49]: We reuse inductor in the next stacked PR.

**Review by EikanWang — APPROVED** [review]

**Comment by EikanWang** [comment]
@etaf , please provide some detailed descriptions for your PR to streamline the review.

**Comment by etaf** [comment]
> @etaf , please provide some detailed descriptions for your PR to streamline the review.

Thanks for the reminder, I've detailed, please have a review.

**Review by jansel — APPROVED** [review]

**Review by desertfire — APPROVED** [review]

File: test/inductor/test_memory_planning.py [hunk 3]
```diff
@@ -116,4 +117,5 @@ def test_abi_compatible(self):
 
 
 if __name__ == "__main__":
-    run_tests()
+    if HAS_CUDA:
```
[the remark this answers is not in the document]
Comment (etaf) [reply test/inductor/test_memory_planning.py:120]: Thanks for your review, if no one picks this up, we can do it at our next PR.

**Pull request closed without merging**